File: output/cpp/tests/04_native_code.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>
#include <cmath>

int helper_function(int x) {
    return x * 2;
}

class NativeCodeFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    NativeCodeFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class NativeCodeFrameContext {
public:
    NativeCodeFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    NativeCodeFrameContext(NativeCodeFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class NativeCodeCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<NativeCodeFrameEvent> forward_event;
    std::unique_ptr<NativeCodeCompartment> parent_compartment;

    explicit NativeCodeCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<NativeCodeCompartment> clone() const {
        auto c = std::make_unique<NativeCodeCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class NativeCode {
private:
    std::vector<std::unique_ptr<NativeCodeCompartment>> _state_stack;
    std::unique_ptr<NativeCodeCompartment> __compartment;
    std::unique_ptr<NativeCodeCompartment> __next_compartment;
    std::vector<NativeCodeFrameContext> _context_stack;

    void __kernel(NativeCodeFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            NativeCodeFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                NativeCodeFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    NativeCodeFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(NativeCodeFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Active") {
            _state_Active(__e);
        }
    }

    void __transition(std::unique_ptr<NativeCodeCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Active(NativeCodeFrameEvent& __e) {
        if (__e._message == "compute") {
            auto value = std::any_cast<int>(__e._parameters.at("value"));
            int temp = value + 10;
            int result = helper_function(temp);
            printf("Computed: %d -> %d\n", value, result);
            _context_stack.back()._return = std::any(result);
            return;;
        } else if (__e._message == "use_math") {
            double result = sqrt(16.0) + M_PI;
            printf("Math result: %f\n", result);
            _context_stack.back()._return = std::any(result);
            return;;
        }
    }

public:
    NativeCode() {
        __compartment = std::make_unique<NativeCodeCompartment>("Active");
        NativeCodeFrameEvent __frame_event("$>");
        NativeCodeFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    int compute(int value) {
        std::unordered_map<std::string, std::any> __params;
        __params["value"] = value;
        NativeCodeFrameEvent __e("compute", std::move(__params));
        NativeCodeFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    double use_math() {
        NativeCodeFrameEvent __e("use_math");
        NativeCodeFrameContext __ctx(std::move(__e), std::any(double()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<double>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 04: Native Code Preservation ===\n");
    NativeCode s;

    int result = s.compute(5);
    int expected = (5 + 10) * 2;
    if (result != expected) {
        printf("FAIL: Expected %d, got %d\n", expected, result);
        assert(false);
    }
    printf("compute(5) = %d\n", result);

    double mathResult = s.use_math();
    double expectedMath = sqrt(16.0) + M_PI;
    if (fabs(mathResult - expectedMath) >= 0.001) {
        printf("FAIL: Expected ~%f, got %f\n", expectedMath, mathResult);
        assert(false);
    }
    printf("use_math() = %f\n", mathResult);

    printf("PASS: Native code preservation works correctly\n");
    return 0;
}

File: output/go/tests/04_native_code.go

package main

import (
	"fmt"
	"math"
	"os"
)

func helperFunction(x int) int {
	// Native helper function defined before system
	return x * 2
}

type NativeCodeFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type NativeCodeFrameContext struct {
    _event  NativeCodeFrameEvent
    _return any
    _data   map[string]any
}

type NativeCodeCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *NativeCodeFrameEvent
    parentCompartment *NativeCodeCompartment
}

func newNativeCodeCompartment(state string) *NativeCodeCompartment {
    return &NativeCodeCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *NativeCodeCompartment) copy() *NativeCodeCompartment {
    nc := &NativeCodeCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type NativeCode struct {
    _state_stack []*NativeCodeCompartment
    __compartment *NativeCodeCompartment
    __next_compartment *NativeCodeCompartment
    _context_stack []NativeCodeFrameContext
}

func NewNativeCode() *NativeCode {
    s := &NativeCode{}
    s._state_stack = make([]*NativeCodeCompartment, 0)
    s._context_stack = make([]NativeCodeFrameContext, 0)
    s.__compartment = newNativeCodeCompartment("Active")
    s.__next_compartment = nil
    __frame_event := NativeCodeFrameEvent{_message: "$>", _parameters: nil}
    __ctx := NativeCodeFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *NativeCode) __kernel(__e *NativeCodeFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &NativeCodeFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &NativeCodeFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &NativeCodeFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *NativeCode) __router(__e *NativeCodeFrameEvent) {
    switch s.__compartment.state {
    case "Active":
        s._state_Active(__e)
    }
}

func (s *NativeCode) __transition(next *NativeCodeCompartment) {
    s.__next_compartment = next
}

func (s *NativeCode) Compute(value int) int {
    __params := map[string]any{
        "value": value,
    }
    __e := NativeCodeFrameEvent{_message: "Compute", _parameters: __params}
    __ctx := NativeCodeFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *NativeCode) UseMath() float64 {
    __e := NativeCodeFrameEvent{_message: "UseMath"}
    __ctx := NativeCodeFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result float64
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(float64) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *NativeCode) _state_Active(__e *NativeCodeFrameEvent) {
    if __e._message == "Compute" {
        value := __e._parameters["value"].(int)
        _ = value
        // Native code with local variables
        temp := value + 10
        result := helperFunction(temp)
        fmt.Printf("Computed: %d -> %d\n", value, result)
        s._context_stack[len(s._context_stack)-1]._return = result
        return
    } else if __e._message == "UseMath" {
        // Using math module
        result := math.Sqrt(16) + math.Pi
        fmt.Printf("Math result: %f\n", result)
        s._context_stack[len(s._context_stack)-1]._return = result
        return
    }
}

func main() {
	fmt.Println("=== Test 04: Native Code Preservation ===")
	sm := NewNativeCode()

	// Test native code in handler with helper function
	result := sm.Compute(5)
	expected := (5 + 10) * 2 // 30
	if result != expected {
		fmt.Printf("FAIL: Expected %d, got %d\n", expected, result)
		os.Exit(1)
	}
	fmt.Printf("Compute(5) = %d\n", result)

	// Test math module usage
	mathResult := sm.UseMath()
	expectedMath := math.Sqrt(16) + math.Pi
	if math.Abs(mathResult-expectedMath) >= 0.001 {
		fmt.Printf("FAIL: Expected ~%f, got %f\n", expectedMath, mathResult)
		os.Exit(1)
	}
	fmt.Printf("UseMath() = %f\n", mathResult)

	fmt.Println("PASS: Native code preservation works correctly")
}
